### DISCUSSION TITLE
Future of tB

### DISCUSSION THREAD

**Discussion by uwbwsvd** [twinbasic/twinbasic#789] [General]
What is the real advantage of tB (VB) if it is not just about not having to learn a new language or concept? How would it fit into a server, web app based structure? Or is it just a special tool for an old-timer that I want to continue using.
I think Microsoft, for example, didn't invent Net just because a few people were bored for a long time.

Perhaps there is already a listing of advantages, that I have overseen. So excuse me.
Advantages for me:
1) I know VBA (no new learning)
2) tB(VB) seems to be very easy to use
3) Creating stand alone exe (so far as I know)

**Comment by EduardoVB** [comment]
This is of course my opinion:

BASIC is one of the most human friendly, easy to learn and easy to work with programming languages.
VB was an evolution of that.
tB is an evolution of VB.
And coming from there, its main goal is to have full compatibility with VB6 and VBA.
But that is the first stage.
After that, it will (hopefully) cover web development and in my sight, it will gradually replace other programming languages like JavaScript and Python.

> I think Microsoft, for example, didn't invent Net just because a few people were bored for a long time.

No, they did it because they feared Java.

BTW: this "issue" should be moved to discussions. Issues are intended for bug reports and feature requests.

**Reply by YogiYang** [reply]
> BASIC is one of the most human friendly, easy to learn and easy to work with programming languages.

I would suggest you to check out [LiveCode](https://livecode.com/). It is a mature cross platform programming tool with a visual IDE and it supports all platforms. Its programming language is plain old English and quite verbos!!

**Comment by uwbwsvd** [comment]
> No, they did it because they feared Java.
What make it so alarming?

**Reply by Greedquest** [reply]
One reason: Java can run on any OS (or the web) because of the JVM (virtual machine running java bytecode). .Net and especially the C#language were developed partly to keep desktop developers attached to Windows (of course Microsoft wants developers to use Windows and write good software for Windows in preference to other operating systems since this drives sales of their own products).

**Comment by EduardoVB** [comment]
> > What make it so alarming?

Better ask them. 
But I think they perceived it as it was going to be a revolution or something like that, and because it also promised independence from the OS (would run on Windows/Linux/Mac/Etc). And it would be a threat to Windows.
So they thought better we make the same, so we can control it and the direction it takes.

**Comment by WaynePhillipsEA** [comment]
The BASIC language itself is so easy to understand, and yet can be so powerful at the same time.  Whether you're a beginner programmer or a seasoned professional, you can easily pickup VB6/tB and be creating things in no time at all.

For me, one of the big advantages that twinBASIC has over many other languages is that there are no massive runtime library dependencies.   Much like VB6, you can create small self-contained applications that can be distributed without having to install anything other than the EXE on the end-user machines.

I also feel that 64-bit support is also going to be important going forward, as I certainly expect there will be a future where 32-bit applications are not supported in newer versions of Windows.

With WebView2 and a built-in high-performance web-server just around the corner, there's a lot to be looking forward to with twinBASIC.  These new features will open up brand new possibilities.

**Reply by Greedquest** [reply]
@WaynePhillipsEA I'm slightly confused about the "no massive runtime library dependencies” point. This feels like an implementation detail of your (fantastic) compiler rather than something intrinsic to twinBASIC the language. I mean surely there's a tradeoff here, does the step of purging extra unused functions not just increase compilation time and complexity, to the extent that many languages just bundle a load of extra functionality in a standard library that doesn't need to be compiled every time. I just think that it's no coincidence that this is an uncommon feature of compilers and I'm wondering if there's something special about tB that permits it. Or are you just a brilliant genius (I could buy this theory too)?

**Reply by WaynePhillipsEA** [reply]
I'm not a.NET fan, but I do believe with. NET core they introduced a 'self-contained' build option, which is somewhat similar.

One major advantage is that you are not relying on the version of an installed runtime affecting your app in any way.  Your app contains everything it needs (apart from OS dependencies of course), and so when searching for the source of errors/problems, there is much less surface area to cover.

Of course the disadvantage is that the final build size of your executables is a bit larger.  Compilation time isn't really affected because the standard libraries are pre-compiled, with parts of them getting linked in to your EXEs/DLLs as necessary.    The custom built tB linker is just pretty smart, that's all :)

**Reply by Greedquest** [reply]
Oh yeah I see why you might want to have a standalone exe. I was more wondering why some languages bundle a large standard library in the exe rather than just pruning it in the way you do to purge unused methods (I think Go does this). Surely there is some tradeoff? Maybe just the tB compiler is smarter as you say.

Edit: it was Go with the big bundled standard library in every exe

![image](https://user-images.githubusercontent.com/25348920/162582261-f10356e0-39f3-4562-a3c4-a4a111b146bd.png)

https://nolongerset.com/twinbasic-runtime-requirements/

**Comment by gregse99** [comment]
I’d like to offer an “outsider’s” perspective on tB.  I do not have a computer science degree and I am not a professional programmer.  But as I have delved into the world of programming more and more for my job (buy-side investor) several things have sort of shocked me about the developer community - and I think one of them is very relevant to tB.  First, the big positive: the developer community is simply one of the most generous communities I’ve ever encountered.  The willingness of people to share their knowledge - even with a non-developer like myself - is humbling.  But there is one thing that has really shocked me on the negative side: I’ve rarely encountered a community that is so pervaded by utterly mindless dogmatism.  The number of ridiculous “holy wars” - with people almost mindlessly clinging to their own technology stack, programming paradigms, preferred languages, etc. is really quite surprising coming from a group of people who are ostensibly smarter than average.  Why do I mention this?  Because I think the clear bias against VbX (it almost seems like it’s “cool” to trash VbX - it always gets the “most dreaded language” award in the Stack Overflow survey) is absolutely insane.  Who decided that it was an advancement to make programmers use brackets rather than simply express in plain language (such as in Python of Basic) their instructions?  To me, every C-derived language seems like an absolute mindless step backwards in terms of the programmer’s ease of expressing the intention and logic of his or her instructions.  At some point won’t people realize that high-level languages should literally be as high-level as possible?  Where anyone can easily learn to instruct a computer in more or less plain english?  If I am right then perhaps computer science will ultimately circle back on itself and embrace basic and other languages that use common-sense syntax.  I think that we are already seeing such a “meta” trend with the explosion of Python - perhaps the closest thing to Basic I have seen (and Python doesn’t even have strong typing like VbX - which strikes me as an absolutely insane omission).  I guess my point is that if there is a meta-trend going on here (toward the “democratization” of coding by using more plain english) then twinBasic could eventually have a HUGE future.  People just need to get over their dogmatic views that VbX is not “cool”.

**Comment by uwbwsvd** [comment]
An analogy.
As I learned tractor driving as a young boy a traktor look like this:
 <img src="https://user-images.githubusercontent.com/41230018/162558485-25ecd101-bcbc-4aef-bd6b-4fcf8d1034d3.png" width="600" height="400">
Now this machine look like this:
![grafik](https://user-images.githubusercontent.com/41230018/162558546-1ae0e35a-3286-4dd5-b470-40d46f851ef9.png)
![grafik](https://user-images.githubusercontent.com/41230018/162558853-a0c76129-2d22-4caf-aff1-4c0187bcc15a.png)

Why? Be faster, be more confortable, be stronger, be more efficient.
Which tractor do you think is easier to learn to drive? 
For some goals you have to learn komplex systems.

**Reply by EduardoVB** [reply]
> For some goals you have to learn komplex systems.

I think you are a bit confused.

Complex does not mean better. In fact it is quite the opposite.
Simpler is better.

This subject has been discussing long already, perhaps not here, but in other places.
Basically there are a couple of reasons why programming involuted in last two decades, of course it is my point of view and understanding.

One is that some programmers want to be an elite, and don't want everybody programming. Some of this kind of programmers had power in corporations to make decisions and to influence trends.
They want programming to be only for a kind of elite that is able to read hieroglyphics.

Another reason is that MS wanted to carry everybody to .NET. They seem to have thought that programmers were an asset that they had and could use as they wanted.
The reason why they wanted to push .NET were already discussed above, at least some of them.

Those reasons combine, because at MS there were programmers, software architects, people of the field.
Corporations buy the press to set trends, the trends that they want. So it is not hard to understand the bad press of VBx.
What many people think is based on what other say, it is not like everybody has an independent and conscientiously opinion based on facts. They just repeat, follow trends.

So that is mostly based on lies.
Yes, there are some things that can be criticized of VBx, like the lack inheritance, not multi-threaded, and some other shortcomings. But the reality is that are more the advantages than the disadvantages. And tB will evolve regarding these shortcomings.

So, what's your point? If you like complex languages you need to look to other languages. I suggest Malbolge.
If you believe everything that you read and follow the mass, then wait until tB become a dominant language.

**Reply by uwbwsvd** [reply]
I work with Access. We have projects that call for SQL server. We need to collect data (we are doing this with Excel). But a smart phone app would be better. At the moment I can't see where TB can help.

**Comment by Greedquest** [comment]
VBA hobbyist p.o.v here. As I've stated elsewhere, I believe VBA will continue to be relevant as long as it is so well integrated into MS Office. The language and editor have some features that make it pretty good at capitalising on this integration and well suited to MS Office users:
 - verbose, human readable (therefore beginner/non-programmer friendly) syntax
 - In my experience, a very good debugging experience, amazingly good considering how old it is. I much prefer it to python in this regard.
 - macro recorder makes it very accessible and discoverable.
 - no setup beyond installing office, code is copy-pasteable
 - the Variant type
 - etc.

So VBA has a place in my opinion. But what about the subject of your question, tB?

Well I don't know whether I see it as a VBA replacement (in the same way it can replace VB6). I think replacing VBA, you are always a bit at the whim of Microsoft. They would upset a lot of people replacing VBA with typescript entirely for example, but they might anyway. My opinion may change on this depending on whether the new tB editor can integrate equally well into Office and supports easy (non-admin or unskilled) install - That may allow users to ignore Microsoft's direction and continue to use VBA/tB.

However what tB does a fantastic job of already is in _supplementing_ VBA. Particularly 2 ways:
 - Writing libraries to reference from VBA. I've tried writing COM libraries in python and it sucks for many reasons (no intellisense, tricky setting up COM server, nearly impossible to debug, ...) . tB is based on the same COM architecture as VBA and integrates seamlessly.
 - COM automation tasks. Again, python is pretty good at automating office applications but not when it needs to interact with the COM API. I just migrated some command line app for automating Excel from python to tB and it's half the code, twice the speed and much more idiomatic since Excel was designed to be called from COM languages like tB.

Basically I subscribe to the vision of tB as the best language for COM libraries and automation. More powerful than VBA, but more idiomatic and with appropriate abstraction than, say, C++ which provides more flexibility than necessary.

Final note - Although I'm not a VB6 programmer I do see tB as replacing it in legacy codebases (appeal of 64 bit, modern editor improves developer experience, more powerful compiler, multi-plarform support). That said, I don't think the appeal of the syntax is enough to take market share away from other modern and equally feature rich automation languages, the ones that stole users away from VB6, but I'd like to be proven wrong (you might get a few who choose tB over another language because they already know the VBA syntax, but I don't know how many that will be).

**Comment by uwbwsvd** [comment]
https://github.com/WaynePhillipsEA/twinbasic/discussions/791#discussioncomment-2571149

Of course, there are plenty of wishes for the programming language.  From my point of view, it is not clear whether tB will be successful in the sense of being widely usable. There is some evidence that it can be. From my point of view, a clearly structured roadmap is of great importance in order to make the best use of the available resources (obviously only Wayne, unfortunately).
In the current state of the project, I don't think it's suitable for real beginners (I'm also a VB beginner). Especially because I can't be sure that errors are mine in programming or perhaps those of the development environment.
At the moment I see the previous VB programmers who would like to continue their projects in tB as real beneficiaries of tB. Therefore, these people should also have a significant say in the Raodmap.
